src/hatch/python/resolve.py

```diff
@@ -55,43 +55,36 @@ def archive_name(self) -> str:
         return self.source.rsplit("/", 1)[-1]
 
     def unpack(self, archive: Path, directory: Path) -> None:
+        # zip
         if self.source.endswith(".zip"):
             import zipfile
 
             with zipfile.ZipFile(archive, "r") as zf:
                 zf.extractall(directory)
-        elif self.source.endswith((".tar.gz", ".tgz")):
-            import tarfile
+            return
 
-            with tarfile.open(archive, "r:gz") as tf:
-                if sys.version_info[:2] >= (3, 12):
-                    tf.extractall(directory, filter="data")
-                else:
-                    tf.extractall(directory)  # noqa: S202
-        elif self.source.endswith((".tar.bz2", ".bz2")):
+        # tar
+        if sys.version_info >= (3, 14):
             import tarfile
+        else:
+            # for zstd support (introduced in Python 3.14)
+            # and filter kwarg (introduced in Python 3.12)
+            from backports.zstd import tarfile
 
-            with tarfile.open(archive, "r:bz2") as tf:
-                if sys.version_info[:2] >= (3, 12):
-                    tf.extractall(directory, filter="data")
-                else:
-                    tf.extractall(directory)  # noqa: S202
+        mode: Literal["r:gz", "r:bz2", "r:zst"]
+        if self.source.endswith((".tar.gz", ".tgz")):
+            mode = "r:gz"
+        elif self.source.endswith((".tar.bz2", ".bz2")):
+            mode = "r:bz2"
         elif self.source.endswith((".tar.zst", ".tar.zstd")):
-            import tarfile
-
-            import zstandard
-
-            with open(archive, "rb") as ifh:
-                dctx = zstandard.ZstdDecompressor()
-                with dctx.stream_reader(ifh) as reader, tarfile.open(mode="r|", fileobj=reader) as tf:
-                    if sys.version_info[:2] >= (3, 12):
-                        tf.extractall(directory, filter="data")
-                    else:
-                        tf.extractall(directory)  # noqa: S202
+            mode = "r:zst"
         else:
             message = f"Unknown archive type: {archive}"
             raise ValueError(message)
 
+        with tarfile.open(archive, mode) as tf:
```

Comment:
Defaulting to a specific format seems structurally odd to me. Please instead make a tarfile+zstandard extraction helper function, or method prefixed by double underscores, whose definition depends on the Python version that will merely be called inside the associated branch here.

Reply:
This appears to address what ofek was asking. @ofek I am going to go ahead and get this merged. If there are any additional changes let me know and I can try to help address them before our release.